<a href="https://colab.research.google.com/github/ITES-Hub/SchoolDev_Technologies_Practical_AI_Assignment/blob/main/Assignment_1_NewEmp_Dataset_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**SchoolDev_Technologies_Pratical_AI_Assignment**
##Full Data Cleaning of NewEmp Dataset

Student_Name: **Ebenezer Oriade**

**Task Instructions:**

Giving the above dataset -NEWEMP.xlsx

 You are required to carry out a full data cleaning exercise

+ Check for Data inconsistency

+ Duplicates

+ Replace Missing values with mean values

+ Outliers Removal or cap them


##Excutive Summary:

This is a comprehensive data cleaning project for the NewEmp dataset.

It covers loading data from Excel, handling missing 'Age' values via mean imputation, identifying and removing duplicate records, and detecting and removing outliers in 'Salary' using the IQR method.

The notebook ensures a clean, analysis-ready dataset.

###**Step_Wise Approach:**

Step 1: Load the Dataset with pandas

In [43]:
import pandas as pd
import numpy as np

In [44]:
data = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/NewEmp.xlsx')
data.head(6)

,Emp_id,Age,Department,Salary
0,PID/729,41.0,Sales,340
1,PID/603,49.0,Research & Development,3470
2,PID/4047,37.0,Research & Development,8900
3,PID/4047,37.0,Research & Development,8900
4,PID/5538,37.0,Research & Development,4500
5,PID/568,NaN,Research & Development,8730


In [45]:
data.shape

(1473, 4)

In [46]:
data.describe()

,Age,Salary
count,1465.000000,1.473000e+03
mean,36.948805,7.050665e+03
std,9.123289,3.205405e+04
min,18.000000,3.400000e+02
25%,30.000000,4.860000e+03
50%,36.000000,6.280000e+03
75%,43.000000,7.590000e+03
max,60.000000,1.234890e+06


In [47]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1473 entries, 0 to 1472
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Emp_id      1473 non-null   object 
 1   Age         1465 non-null   float64
 2   Department  1473 non-null   object 
 3   Salary      1473 non-null   int64  
dtypes: float64(1), int64(1), object(2)
memory usage: 46.2+ KB


In [48]:
data['Age']

,Age
0,41.0
1,49.0
2,37.0
3,37.0
4,37.0
...,...
1468,36.0
1469,39.0
1470,NaN
1471,49.0


In [49]:
data['Age'].isna().sum()

np.int64(8)

In [50]:
mean_age = data['Age'].mean()
data['Age'].fillna(mean_age, inplace=True)

print(f"Missing values in 'Age' after imputation: {data['Age'].isna().sum()}")

Missing values in 'Age' after imputation: 0


In [51]:
display(data.head(6))

,Emp_id,Age,Department,Salary
0,PID/729,41.000000,Sales,340
1,PID/603,49.000000,Research & Development,3470
2,PID/4047,37.000000,Research & Development,8900
3,PID/4047,37.000000,Research & Development,8900
4,PID/5538,37.000000,Research & Development,4500
5,PID/568,36.948805,Research & Development,8730


In [52]:
print(f"Number of duplicate rows: {data.duplicated().sum()}")

Number of duplicate rows: 2


In [53]:
data[data.duplicated(keep=False)]

,Emp_id,Age,Department,Salary
2,PID/4047,37.0,Research & Development,8900
3,PID/4047,37.0,Research & Development,8900
6,PID/1863,27.0,Research & Development,6920
7,PID/1863,27.0,Research & Development,6920


In [54]:
data.drop_duplicates(inplace=True)

In [55]:
data[data.duplicated(keep=False)]

,Emp_id,Age,Department,Salary


In [56]:
numerical_columns = data.select_dtypes(include=np.number).columns
print(numerical_columns)

Index(['Age', 'Salary'], dtype='object')


NB:The above result shows that Age and Salary columns are the only numerical columns upon which otlier detection can be carried out.

In [57]:
print("Outlier Detection using IQR Method:")
for col in numerical_columns:
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = data[(data[col] < lower_bound) | (data[col] > upper_bound)]

    if not outliers.empty:
        print(f"\nColumn: {col}")
        print(f"  Q1: {Q1:.2f}")
        print(f"  Q3: {Q3:.2f}")
        print(f"  IQR: {IQR:.2f}")
        print(f"  Lower Bound: {lower_bound:.2f}")
        print(f"  Upper Bound: {upper_bound:.2f}")
        print(f"  Number of outliers: {len(outliers)}")
        display(outliers)
        print("\n")
    else:
        print(f"No outliers found in column: {col}")

Outlier Detection using IQR Method:
No outliers found in column: Age

Column: Salary
  Q1: 4855.00
  Q3: 7590.00
  IQR: 2735.00
  Lower Bound: 752.50
  Upper Bound: 11692.50
  Number of outliers: 2


,Emp_id,Age,Department,Salary
0,PID/729,41.0,Sales,340
1471,PID/1104,49.0,Sales,1234890


The above result shows the rows and particular columns with outliers.

In [58]:
# Remove outliers from the DataFrame
data.drop(outliers.index, inplace=True)

print(f"Shape of DataFrame after removing outliers: {data.shape}")

Shape of DataFrame after removing outliers: (1469, 4)


### **Re-checking for Outliers after Remova**l
Let's run the outlier detection again to confirm that the outliers have been successfully removed.

In [59]:
print("Outlier Detection using IQR Method (after removal):")
for col in numerical_columns:
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers_after_removal = data[(data[col] < lower_bound) | (data[col] > upper_bound)]

    if not outliers_after_removal.empty:
        print(f"\nColumn: {col}")
        print(f"  Number of outliers: {len(outliers_after_removal)}")
        display(outliers_after_removal)
    else:
        print(f"No outliers found in column: {col} (after removal)")

Outlier Detection using IQR Method (after removal):
No outliers found in column: Age (after removal)
No outliers found in column: Salary (after removal)


**NB:** Outliers can be either removed or capped. I.e., involves limiting the extreme values in your data to a certain threshold. Instead of removing them, values above an upper limit are replaced with that upper limit, and values below a lower limit are replaced with that lower limit.



**Impact on Analysis:**

**Removing:** Can lead to a loss of data, which might reduce the sample size and potentially the statistical power or generalizability of your findings. It can also bias the sample if outliers are removed non-randomly.

**Capping:** Retains all observations, preventing data loss. However, it can distort the distribution of the variable, potentially reducing the variance and making the data appear less spread out than it truly is. It also introduces an artificial upper/lower bound.

###**In Conlusion**

In our case, the two 'Salary' outliers (340 and 1,234,890) seem quite extreme.

The 340 salary might be an error (too low), and 1,234,890 is exceptionally high compared to the rest of the salaries.

In this analysis, the remove option was preferred. However, if this were a real-world scenario, we'd investigate these values more closely.




###**Closing Remark**

Thanks for going through my data cleaning work.

Watchout for more insight from me as I explore more real world data...

**--Ebenezer Oriade**